File: Client.React/src/utils/gameHelpers.ts
import type { Competition, Competitor, EspnScores, Event, TypeName, HomeAway, EspnRecordType } from '../types/espn';
import { toLocalDisplay } from './time';
import type { PickType } from '../types/picks';

export function getWeekFromEspnWeek(week: number, isPostSeason = false) {
  return isPostSeason ? week + 18 : week;
}

export function getWeekName(week: number, isPostSeason = false) {
  if (!isPostSeason) return `Week ${week}`;
  switch (week) {
    case 1:
      return 'Wild Card';
    case 2:
      return 'Divisional Round';
    case 3:
      return 'Conference Championship';
    case 4:
      return 'Super Bowl';
    default:
      throw new Error('Invalid week number');
  }
}

export function getEspnRequiredPicks(week: number, isPostSeason = false) {
  if (!isPostSeason) return 4;
  switch (week) {
    case 1:
    case 2:
      return 3;
    case 3:
      return 2;
    case 4:
    case 5:
      return 1;
    default:
      throw new Error('Invalid week number');
  }
}

const STATUS_FINAL = 0;
const STATUS_HALFTIME = 1;
const STATUS_IN_PROGRESS = 2;
const STATUS_SCHEDULED = 3;

function isStatus(status: TypeName | string | number, code: number, text: string) {
  if (typeof status === 'number') return status === code;
  return status === text;
}

export function displayDetails(competition: Competition): string {
  const status = competition.status.type.name;
  if (isStatus(status, STATUS_SCHEDULED, 'status_scheduled')) {
    return toLocalDisplay(competition.date, {
      weekday: 'short',
      hour: 'numeric',
      minute: '2-digit',
    });
  }
  if (isStatus(status, STATUS_HALFTIME, 'status_halftime')) return 'Half Time';
  if (isStatus(status, STATUS_IN_PROGRESS, 'status_in_progress')) {
    return `Q${competition.status.period} ${competition.status.displayClock}`;
  }
  if (isStatus(status, STATUS_FINAL, 'status_final')) return 'Final';
  return '';
}

export function getCompetitionFromHomeAwayAbbr(home: string, away: string, scores: EspnScores) {
  for (const scoreEvent of scores.events ?? []) {
    for (const competition of scoreEvent.competitions) {
      if (getHomeTeamAbbr(competition) === home && getAwayTeamAbbr(competition) === away) {
        return competition;
      }
    }
  }
  throw new Error('Competition not found');
}

export function getAwayTeamAbbr(competition: Competition) {
  return getTeamAbbr(getAwayTeam(competition));
}

export function getHomeTeamAbbr(competition: Competition) {
  return getTeamAbbr(getHomeTeam(competition));
}

export function getAwayTeam(competition: Competition) {
  const found = competition.competitors.find((c) => isHomeAway(c.homeAway, 'away'));
  if (!found) throw new Error('Away team missing');
  return found;
}

export function getHomeTeam(competition: Competition) {
  const found = competition.competitors.find((c) => isHomeAway(c.homeAway, 'home'));
  if (!found) throw new Error('Home team missing');
  return found;
}

export function getAwayTeamLogo(competition: Competition) {
  return getTeamLogo(getAwayTeamAbbr(competition));
}

export function getHomeTeamLogo(competition: Competition) {
  return getTeamLogo(getHomeTeamAbbr(competition));
}

export function getTeamLogo(teamAbbr: string) {
  return `/Icons/Teams/${teamAbbr.toLowerCase()}.png`;
}

export function getTeamScore(competitor: Competitor): number {
  const raw = competitor.score as unknown;
  if (typeof raw === 'number') return raw;
  const parsed = Number(raw);
  return Number.isFinite(parsed) ? parsed : 0;
}

export function getHomeTeamScore(competition: Competition) {
  return getTeamScore(getHomeTeam(competition));
}

export function getAwayTeamScore(competition: Competition) {
  return getTeamScore(getAwayTeam(competition));
}

export function isGameStarted(competition: Competition) {
  return !isStatus(competition.status.type.name, STATUS_SCHEDULED, 'status_scheduled');
}

export function isGameOver(competition: Competition) {
  return isStatus(competition.status.type.name, STATUS_FINAL, 'status_final');
}

export function getTeamAbbr(competitor: Competitor | undefined) {
  return competitor?.team?.abbreviation ?? '';
}

export function getTeamRecord(competitor: Competitor) {
  return competitor.records?.find((r) => isRecordType(r.type, 'total'))?.summary;
}

export function getDownDistance(competition: Competition) {
  return competition.situation?.downDistanceText;
}

export function isRedZone(competition: Competition) {
  return Boolean(competition.situation?.isRedZone);
}

export function getPossessionTeamAbbr(competition: Competition) {
  const possessionId = competition.situation?.possession;
  if (!possessionId) return null;
  const team = competition.competitors.find((c) => c.id === possessionId);
  return team?.team?.abbreviation ?? null;
}

export function isHalfTime(competition: Competition) {
  return isStatus(competition.status.type.name, STATUS_HALFTIME, 'status_halftime');
}

export function hasPossession(competition: Competition, teamAbbr: string) {
  const possession = getPossessionTeamAbbr(competition);
  return possession === teamAbbr;
}

export function isAfterKickoff(competition: Competition, now: Date = new Date()): boolean {
  return now >= new Date(competition.date);
}

export function shouldShowGamePicks(competition: Competition) {
  return isGameStarted(competition) || isAfterKickoff(competition);
}

export function isPostSeason(scores: EspnScores | null) {
  if (!scores?.season) return false;
  return scores.season.type === 3;
}

export function getScoreEvents(scores: EspnScores | null | undefined): Event[] {
  return scores?.events ?? [];
}

export function getPickLabel(pickType: PickType) {
  return pickType === 'Spread' ? 'Spread' : pickType;
}

function isHomeAway(value: HomeAway, expected: 'home' | 'away') {
  if (typeof value === 'number') {
    return expected === 'away' ? value === 0 : value === 1;
  }
  return value === expected;
}

function isRecordType(value: EspnRecordType, expected: 'total' | 'road' | 'home') {
  if (typeof value === 'number') {
    if (expected === 'total') return value === 0;
    if (expected === 'road') return value === 1;
    return value === 2;
  }
  return value === expected;
}

File: Shared/Helpers/GameHelpers.cs
using FourPlayWebApp.Shared.Models;

namespace FourPlayWebApp.Shared.Helpers;

public static class GameHelpers {
    // Shared by NflLiveScoreFetcher and CfbLiveScoreFetcher — defense in depth on top of each
    // fetcher's own dates= query param (frizat-11t): don't fully trust ESPN to honor a date range
    // perfectly either (e.g. a timezone-boundary edge case on a late West Coast kickoff). One
    // implementation instead of drifting per-sport copies of the same boundary check.
    public static Event[] FilterEventsToDateWindow(IEnumerable<Event> events, DateTime start, DateTime end) =>
        events.Where(e => {
            var comp = e.Competitions.FirstOrDefault();
            return comp is not null && comp.Date.Date >= start.Date && comp.Date.Date <= end.Date;
        }).ToArray();

    public static EspnScores WithEvents(EspnScores source, Event[] events) => new() {
        Leagues = source.Leagues,
        Season  = source.Season,
        Week    = source.Week,
        Events  = events,
    };

    public static bool IsPastNoonCst
    {
        get
        {
            var nowUtc = PageTimeProvider.UtcNow;
            var nowCst = TimeZoneHelpers.ConvertTimeToCst(nowUtc);

            // Sunday noon CST
            var sundayNoon = nowCst.Date.AddDays(-(int)nowCst.DayOfWeek).AddHours(12);

            // Monday 11:50 PM CST
            var monday = sundayNoon.AddDays(1);
            var monday1150Pm = monday.Date.AddHours(23).AddMinutes(50);

            // True if current time is between Sunday noon and Monday 11:50 PM
            return nowCst >= sundayNoon && nowCst <= monday1150Pm;
        }
    }

    public static bool IsNoonCst
    {
        get {
            var until = UntilNoonCst();
            if (until is null || until.Value.TotalMinutes <= 0)
                return true;
            return false;
        }
    }

    private static TimeSpan? UntilNoonCst() {
        var nowUtc = PageTimeProvider.UtcNow;
        var nowCst = TimeZoneHelpers.ConvertTimeToCst(nowUtc);

        // Figure out the upcoming Sunday noon in CST
        int daysUntilSunday = ((int)DayOfWeek.Sunday - (int)nowCst.DayOfWeek + 7) % 7;
        var nextSundayNoon = nowCst.Date.AddDays(daysUntilSunday).AddHours(12);

        // If it's already past Sunday noon this week, move to next week
        if (nowCst >= nextSundayNoon)
            return null;

        // Convert target back to UTC so subtraction is safe
        return nextSundayNoon - nowCst;
    }
    public static string? DaysHoursMinutesUntilNoonCst() {
        var span = UntilNoonCst();
        return span is null ? null : $"{span.Value.Days}d {span.Value.Hours}h {span.Value.Minutes}m";
    }
    // TODO(docs/ESPNProBowl.md): verify against a real ESPN response once the 2026 postseason
    // bracket actually exists (~Jan 2027) — the NFL discontinued the Pro Bowl GAME starting with
    // the 2026 season (announced 2026-08-26), so ESPN's postseason week numbering is expected to
    // close the old week-4 gap from here on. See docs/ESPNProBowl.md for the full writeup; update
    // this value if this guess turns out wrong. Public (not the private it was originally) so
    // ESPNApiService.FixEspnProbBowlWeek can gate its own identical quirk-correction on the same
    // threshold instead of carrying an independent, unconditional copy (frizat-4k9) — this and
    // GetWeekFromEspnWeek below are the only two places this threshold is allowed to live.
    public const int LastSeasonEspnSkippedProBowlWeek = 2025;

    // Through the 2025 season, ESPN skipped postseason week 4 for the Pro Bowl — Super Bowl was
    // raw ESPN week 5 (one slot past Conference Championship=3), not the 4th round it actually is.
    // Gated by SEASON, not by the raw week value: for 2026+ we deliberately do NOT apply the old
    // week==5 special case, so if that guess is wrong, a real ESPN week=5 postseason response for
    // 2026+ maps to a nonexistent WeekId (23) and fails loudly (no matching config/DB rows found)
    // instead of silently coinciding with the right answer by accident.
    public static int GetWeekFromEspnWeek(long week, int season, bool isPostSeason = false) {
        if (!isPostSeason) return (int)week;
        if (season <= LastSeasonEspnSkippedProBowlWeek && week == 5) return 22;
        return (int)(week + 18);
    }
    public static string GetWeekName(long week, bool isPostSeason = false) {
        if (!isPostSeason) {
            return $"Week {week}";
        }
        return week switch {
            1 => "Wild Card",
            2 => "Divisional Round",
            3 => "Conference Championship",
            4 => "Super Bowl",
            _ => throw new ArgumentException("Invalid week number")
        };
    }
    public static int GetWeekFromName(string weekName, bool isPostSeason = false) {
        if (!isPostSeason) {
            return int.Parse(weekName.Replace("Week ", ""));
        }

        return weekName switch {
            "Wild Card" => 1,
            "Divisional Round" => 2,
            "Conference Championship" => 3,
            "Super Bowl" => 4,
            _ => throw new ArgumentException("Invalid week name")
        };
    }

    // ESPN resets postseason weeks to 1-5: 1=Wild Card, 2=Divisional, 3=Conf., 4=Pro Bowl, 5=Super Bowl
    public static int GetEspnRequiredPicks(long week, bool isPostSeason = false) {
        if (!isPostSeason) return 4;
        return week switch {
            1 => 3,
            2 => 3,
            3 => 2,
            4 => 1,
            5 => 1, // ESPN Treats Post Season Week 4 as the Pro Bowl - this sucks
            _ => throw new ArgumentException("Invalid week number")
        };
    }

    public static int GetRequiredPicks(long week) {
        if (week < 19) {
            return 4;
        }
        return week switch {
            19 => 3,
            20 => 3,
            21 => 2,
            22 => 1,
            _ => throw new ArgumentException("Invalid week number")
        };
    }

    // 18-slate system: Standard(1-14)=4, NFLDivisional(15-16)=3, NFLConference(17)=2, NFLSuperBowl(18)=1
    public static int GetCfbRequiredPicks(int slateNumber) => slateNumber switch {
        <= 14 => 4,
        <= 16 => 3,
        17 => 2,
        _ => 1,
    };

    public static string? DisplayDetails(Competition competition) {
        return competition.Status.Type.Name switch {
            TypeName.StatusScheduled => TimeZoneHelpers.ConvertTimeToCst(competition.Date.DateTime)
                .ToString("ddd, MMMM dd hh:mm tt"),
            TypeName.StatusHalftime => "Half Time",
            TypeName.StatusInProgress or TypeName.StatusEndPeriod => "In Progress",
            TypeName.StatusFinal => "Final",
            _ => "          "
        };
    }
    public static Competition GetCompetitionFromHomeAwayAbbr(string homeTeamAbbr, string awayTeamAbbr, EspnScores scores) {
        foreach (var scoreEvent in scores.Events) {
            foreach (var competition in scoreEvent.Competitions) {
                if (GetHomeTeamAbbr(competition) == homeTeamAbbr && GetAwayTeamAbbr(competition) == awayTeamAbbr) {
                    return competition;
                }
            }
        }
        throw new ArgumentException("Competition not found");
    }

    public static string GetAwayTeamAbbr(Competition competition) => GetTeamAbbr(GetAwayTeam(competition));
    public static string GetHomeTeamAbbr(Competition competition) => GetTeamAbbr(GetHomeTeam(competition));
    public static Competitor GetAwayTeam(Competition competition) => competition.Competitors.First(x => x.HomeAway == HomeAway.Away);
    public static Competitor GetHomeTeam(Competition competition) => competition.Competitors.First(x => x.HomeAway == HomeAway.Home);
    //public static string GetAwayTeamLogo(Competition competition) => GetAwayTeam(competition).Team.Logo.ToString();
    //public static string GetHomeTeamLogo(Competition competition) => GetHomeTeam(competition).Team.Logo.ToString();
    public static string GetAwayTeamLogo(Competition competition) => GetTeamLogo(GetAwayTeamAbbr(competition));
    public static string GetHomeTeamLogo(Competition competition) => GetTeamLogo(GetHomeTeamAbbr(competition));
    public static string GetTeamLogo(string teamAbbr) => $"Icons/Teams/{teamAbbr.ToLower()}.png";
    public static long GetHomeTeamScore(Competition competition) => GetTeamScore(GetHomeTeam(competition));
    public static long GetAwayTeamScore(Competition competition) => GetTeamScore(GetAwayTeam(competition));
    public static bool IsGameStarted(Competition competition) => competition.Status.Type.Name != TypeName.StatusScheduled;

    // GameTime-based sibling of IsGameStarted(Competition) above — for CfbSpreads/NflSpreads rows,
    // which have no live ESPN status to check. Used by CfbLeaderboardService/LeaderboardService to
    // tell a terminal MissingPicks (every game for the week/slate has already kicked off) apart
    // from a still-open picking window (MissingGameResults) — same GameTime <= now boundary
    // CfbPicksController.StartedTeams already enforces for locking individual picks.
    // frizat-8y6: .All() is vacuously true on an empty sequence — a week/slate with literally no
    // games (spreads not released yet) must never read as "already started" (that flips a
    // genuinely not-yet-open picking window into a terminal MissingPicks loss). Require at least
    // one real game before concluding anything has started.
    public static bool AllGamesStarted(IEnumerable<DateTimeOffset> gameTimes, DateTimeOffset now) {
        var times = gameTimes as ICollection<DateTimeOffset> ?? gameTimes.ToList();
        return times.Count > 0 && times.All(t => t <= now);
    }

    // Shared by LeagueController.AddPicks (NflSpreads) and CfbPicksController.AddPicks
    // (CfbSpreads) — was two byte-identical private methods, one per controller, differing only
    // in the spread row's concrete type. Pure control-table timestamp check, no ESPN dependency.
    public static HashSet<string> StartedTeams<T>(
        IEnumerable<T> spreads, DateTimeOffset now,
        Func<T, DateTimeOffset> gameTime, Func<T, string> homeTeam, Func<T, string> awayTeam) =>
        spreads.Where(s => gameTime(s) <= now).SelectMany(s => new[] { homeTeam(s), awayTeam(s) }).ToHashSet();
    public static bool IsGameOver(Competition competition) => competition.Status.Type.Name == TypeName.StatusFinal;
    public static long GetTeamScore(Competitor competitor) => competitor.Score;
    public static string GetTeamAbbr(Competitor competitor) => competitor.Team.Abbreviation;
    public static string? GetTeamRecord(Competitor competitor) =>
        competitor.Records.FirstOrDefault(x => x.Type == EspnRecordType.Total)?.Summary;

    public static string? GetDownDistance(Competition competition) => competition?.Situation?.DownDistanceText;

    public static bool IsRedZone(Competition competition) {
        if (competition is null)
            return false;
        if (competition.Situation is null)
            return false;
        if (competition.Situation.PossessionText is null)
            return false;
        if (competition.Situation.IsRedZone is null)
            return false;
        return competition.Situation.IsRedZone.Value;
    }

    public static string? GetPossessionTeamAbbr(Competition competition) {
        if (competition.Situation is null)
            return null;
        var possessionId = competition.Situation.Possession;
        if (possessionId is null)
            return null;
        var team = competition.Competitors.FirstOrDefault(x => x.Id == possessionId);
        return GetTeamAbbr(team);
    }
    public static bool IsHalfTime(Competition competition) => competition.Status.Type.Name == TypeName.StatusHalftime;
    public static bool HasPossession(Competition competition, string teamAbbr) {
        var possessionTeamAbbr = GetPossessionTeamAbbr(competition);
        if (possessionTeamAbbr is null)
            return false;
        return possessionTeamAbbr == teamAbbr;
    }


}
